#### Exercici 1. Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

NOTA: Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. Tots els exercicis s'han de fer amb aquest dataset reduït.

In [36]:
import pandas as pd

flights_all_data = pd.read_csv('/Users/juliacorominasfont/Downloads/archive (7)/DelayedFlights.csv')
flights_all_data.shape

(1936758, 30)

In [37]:
#Reducció a 200.000 registres:

flights = flights_all_data.sample(n=200000, random_state=17)
flights.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
534325      1676108  2008      3          16          7   1924.0        1815   
1047777     3502871  2008      6          27          5   1203.0        1155   
1582051     5690590  2008     10           5          7   1625.0        1610   
1284937     4295106  2008      8          21          4   1458.0        1450   
1727359     6437866  2008     11          15          6   1308.0        1225   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
534325    2211.0        2120            AA  ...     8.0    12.0          0   
1047777   1741.0        1730            AA  ...     3.0    24.0          0   
1582051   1723.0        1709            EV  ...     5.0    10.0          0   
1284937   2222.0        2230            WN  ...     4.0    12.0          0   
1727359   1605.0        1507            CO  ...    14.0    40.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
534325                  N         0          10.0           0.0      0.0   
1047777                 N         0           NaN           NaN      NaN   
1582051                 N         0           NaN           NaN      NaN   
1284937                 N         0           NaN           NaN      NaN   
1727359                 N         0           0.0           0.0     24.0   

        SecurityDelay  LateAircraftDelay  
534325            0.0               41.0  
1047777           NaN                NaN  
1582051           NaN                NaN  
1284937           NaN                NaN  
1727359           0.0               34.0  

[5 rows x 30 columns]

El dataset inicial tenia 1936758 files i 30 columnes. Després d'haver-lo reduït conté un total de 200.000 registres i les variables següents:

- Year: 2008 en tots els casos
- Month: d'1 a 12
- DayofMonth: d'1 a 31
- DayOfWeek: d'1 (dilluns) a 7 (diumenge)
- DepTime: hora de sortida (hhmm)
- CRSDepTime: hora de sortida prevista (hhmm)
- ArrTime: hora d'arribada (hhmm)
- CRSArrTime: hora d'arribada prevista (hhmm)
- UniqueCarrier: codi del transportista
- FlightNum: número del vol
- TailNum: número de cua de l'avió
- ActualElapsedTime: temps transcorregut en minuts.. Resultat de sumar les variables: AirTime, TaxiIn i TaxiOut.
- CRSElapsedTime: temps transcorregut previst en minuts..
- AirTime: duració del vol en minuts.
- ArrDelay: Retard respecte l'hora prevista d'arribada, en minuts. Es considera que un avió ha arribat a temps quan el retard és de menys de 15 minuts.
- DepDelay: Retard respecte l'hora de sortida prevista, en minuts.
- Origin: Codi de l'aeroport d'origen.
- Dest: Codi de l'aeroport de destí.
- Distance: Distancia entre el punt d'origen i de destí, en milles.
- TaxiIn: en minuts.
- TaxiOut: en minuts.
- Cancelled: si l'avió es va cancel·lar. 0 quan no es va cancel·lar i 1 en cas que es cancelés.
- CancellationCode: raó de la cancel·lació del vol. A = carrier, B = weather, C = NAS i D = security.
- Diverted: si l'avió es va desviar. 1 en cas que si, i 0 en cas que no. 
- CarrierDelay: en minuts. Retard associat al transportista.
- WeatherDelay: en minuts. Retard associat a condicions meteorològiques.
- NASDelay: en minuts. Retard associat al Sistema Nacional de l'Espai Aèri. 
- SecurityDelay: en minuts. Retard associat a la seguretat.
- LateAircraftDelay: en minuts. Retard associat al retard del mateix avió a l'anterior destació. Aquest efecte es coneix com la propagació del retard.  


A continuació s'eliminen aquelles columnes que no són rellevants per l'estudi, conservant únicament les variables d'interès. 

In [38]:
#Creació d'una nova columna amb la data ['Date'] per poder eliminar les columnes 'Year', 'Month' i 'DayofMonth'

from datetime import datetime
flights['Date'] = flights.apply(lambda row: datetime(row['Year'], row['Month'], row['DayofMonth']), axis=1)

In [39]:
#Eliminació de les columnes que no són d'interès

flights = flights.drop("Unnamed: 0", axis=1) #No aporta informació.
flights = flights.drop("Year", axis=1) #Informació ja inclosa a la columna 'Date'
flights = flights.drop("Month", axis=1) #Informació ja inclosa a la columna 'Date'
flights = flights.drop("DayofMonth", axis=1) #Informació ja inclosa a la columna 'Date'
flights = flights.drop("DepTime", axis=1) #La columna d'interès és la que fa referència a l'hora prevista de sortida. La informació d'aquesta columna ja es pot extreure de la columna DepDelay.
flights = flights.drop("ArrTime", axis=1) #La columna d'interès és la que fa referència a l'hora prevista d'arribada. La informació d'aquesta columna ja es pot extreure de la columna ArrDelay.
flights = flights.drop("ActualElapsedTime", axis=1) #No aporta informació rellevant per l'anàlisis.
flights = flights.drop("CRSElapsedTime", axis=1) #No aporta informació rellevant per l'anàlisis.
flights = flights.drop("TaxiIn", axis=1) #No aporta informació rellevant per l'anàlisis, ja que no permet explicar el retard.
flights = flights.drop("TaxiOut", axis=1) #No aporta informació rellevant per l'anàlisis, ja que no permet explicar el retard.
flights = flights.drop("Diverted", axis=1) #No aporta informació rellevant per l'anàlisis.

In [40]:
flights.head()

DayOfWeek  CRSDepTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
534325           7        1815        2120            AA        848  N494AA   
1047777          5        1155        1730            AA       1414  N624AA   
1582051          7        1610        1709            EV       5262  N879AS   
1284937          4        1450        2230            WN       3571  N728SW   
1727359          6        1225        1507            CO       1493  N67052   

         AirTime  ArrDelay  DepDelay Origin Dest  Distance  Cancelled  \
534325      87.0      51.0      69.0    DFW  ATL       732          0   
1047777    191.0      11.0       8.0    SEA  STL      1710          0   
1582051     43.0      14.0      15.0    TYS  ATL       152          0   
1284937    248.0      -8.0       8.0    LAS  IAD      2066          0   
1727359    123.0      58.0      43.0    MCO  EWR       938          0   

        CancellationCode  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
534325                 N          10.0           0.0       0.0            0.0   
1047777                N           NaN           NaN       NaN            NaN   
1582051                N           NaN           NaN       NaN            NaN   
1284937                N           NaN           NaN       NaN            NaN   
1727359                N           0.0           0.0      24.0            0.0   

         LateAircraftDelay       Date  
534325                41.0 2008-03-16  
1047777                NaN 2008-06-27  
1582051                NaN 2008-10-05  
1284937                NaN 2008-08-21  
1727359               34.0 2008-11-15

In [41]:
flights.shape

(200000, 20)

Després d'eliminar les columnes que no són d'interès per l'anàlisis el dataset té 200.000 registres i 20 columnes.

#### 2.1 Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

In [42]:
flights.describe()

DayOfWeek     CRSDepTime    CRSArrTime     FlightNum  \
count  200000.000000  200000.000000  200000.00000  200000.00000   
mean        3.987955    1466.915465    1634.40038    2179.34677   
std         1.996317     425.442661     464.43067    1940.41706   
min         1.000000       0.000000       0.00000       1.00000   
25%         2.000000    1134.000000    1325.00000     605.00000   
50%         4.000000    1510.000000    1705.00000    1542.00000   
75%         6.000000    1815.000000    2015.00000    3414.00000   
max         7.000000    2359.000000    2400.00000    9740.00000   

             AirTime       ArrDelay       DepDelay       Distance  \
count  199128.000000  199128.000000  200000.000000  200000.000000   
mean      108.258618      42.487084      43.395245     765.544590   
std        68.540561      57.375247      53.933094     573.736573   
min         0.000000     -68.000000       6.000000      31.000000   
25%        58.000000       9.000000      12.000000     340.000000   
50%        90.000000      25.000000      24.000000     607.000000   
75%       137.000000      56.000000      53.000000     997.000000   
max       651.000000    1382.000000    1395.000000    4962.000000   

           Cancelled   CarrierDelay   WeatherDelay       NASDelay  \
count  200000.000000  129094.000000  129094.000000  129094.000000   
mean        0.000315      19.336607       3.738694      15.180651   
std         0.017745      44.252219      21.848971      34.065406   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       2.000000       0.000000       2.000000   
75%         0.000000      21.000000       0.000000      15.000000   
max         1.000000    1382.000000    1070.000000     992.000000   

       SecurityDelay  LateAircraftDelay  
count  129094.000000      129094.000000  
mean        0.087231          25.264381  
std         2.066989          41.986195  
min         0.000000           0.000000  
25%         0.000000           0.000000  
50%         0.000000           8.000000  
75%         0.000000          33.000000  
max       357.000000        1002.000000

Aquí veiem un resum estadístic de les variables d'interès.

Com podem observar, la mitjana de les variables relacionades amb els retards i les seves causes (ArrDelay, DepDelay, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay i LateAircraftDelay) és notablement superior a la seva mediana. Això vol dir que la distribució d'aquestes variables està sesgada cap a la dreta i, per tant, que la mitjana es veu afectada pels valors més alts, tot i haver-ne menys.

El mateix succeeix amb la variable Distance. La seva distribució també està sesgada cap a la dreta, de manera que hi ha algunes rutes que destaquen per ser més llargues que el resta i que augmenten la mitjana.

#### 2.2 Troba quantes dades faltants hi ha per columna.

In [43]:
flights.isnull().sum()

DayOfWeek                0
CRSDepTime               0
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  0
AirTime                872
ArrDelay               872
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
Cancelled                0
CancellationCode         0
CarrierDelay         70906
WeatherDelay         70906
NASDelay             70906
SecurityDelay        70906
LateAircraftDelay    70906
Date                     0
dtype: int64

Les columnes AirTime i ArrDelay tenen 872 valors faltants, que corresponen als casos en els que el vol s'ha cancel·lat. A més a més, les columnes relacionades amb les causes del retard (CarrierDelay, CarrierDelay, NASDelay, SecurityDelay i LateAircraftDelay) tenen 70906, que corresponen als casos en els que no hi ha hagut retard, ja sigui perquè el vol s'ha cancel·lat o perquè ha arribat en hora.

#### 2.3 Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).

In [44]:
flights['Medium_speed'] = flights['Distance'] / flights['AirTime'] #milles per minut

flights['Delay'] = (flights['ArrDelay'] >= 15).astype(int) #Es considera que un avió ha arribat amb retard si aquest és de 15 minuts o més

In [45]:
flights.head()

DayOfWeek  CRSDepTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
534325           7        1815        2120            AA        848  N494AA   
1047777          5        1155        1730            AA       1414  N624AA   
1582051          7        1610        1709            EV       5262  N879AS   
1284937          4        1450        2230            WN       3571  N728SW   
1727359          6        1225        1507            CO       1493  N67052   

         AirTime  ArrDelay  DepDelay Origin  ... Cancelled  CancellationCode  \
534325      87.0      51.0      69.0    DFW  ...         0                 N   
1047777    191.0      11.0       8.0    SEA  ...         0                 N   
1582051     43.0      14.0      15.0    TYS  ...         0                 N   
1284937    248.0      -8.0       8.0    LAS  ...         0                 N   
1727359    123.0      58.0      43.0    MCO  ...         0                 N   

         CarrierDelay WeatherDelay  NASDelay  SecurityDelay  \
534325           10.0          0.0       0.0            0.0   
1047777           NaN          NaN       NaN            NaN   
1582051           NaN          NaN       NaN            NaN   
1284937           NaN          NaN       NaN            NaN   
1727359           0.0          0.0      24.0            0.0   

         LateAircraftDelay       Date  Medium_speed Delay  
534325                41.0 2008-03-16      8.413793     1  
1047777                NaN 2008-06-27      8.952880     0  
1582051                NaN 2008-10-05      3.534884     0  
1284937                NaN 2008-08-21      8.330645     0  
1727359               34.0 2008-11-15      7.626016     1  

[5 rows x 22 columns]

Veiem que ens els casos en els que el retard és inferior a 15 minuts i, per tant, quan Delay = 0, les variables referents a les causes del retard no tenen valors. 

#### 2.4 Fes una taula de les aerolínies amb més endarreriments acumulats.

In [46]:
print('Aerolínies amb més endarreriments acumulats:\n')

delay_minutes_per_carrier = flights.groupby('UniqueCarrier')['ArrDelay'].sum()
delay_minutes_per_carrier.sort_values(ascending = False).head().to_frame()

Aerolínies amb més endarreriments acumulats:



ArrDelay
UniqueCarrier           
WN             1171499.0
AA              923305.0
UA              707163.0
MQ              677499.0
OO              618732.0

Les aerolínies amb més endarreriments acumulats (en minuts) són: WN, AA, UA, MQ i OO.

#### 2.5 Quins són els vols més llargs?

In [47]:
print("Informació dels vols més llargs:\n")

longest_flights = flights.sort_values(by='Distance', ascending=False)
longest_flights[['Distance', 'UniqueCarrier', 'FlightNum', 'TailNum', 'Origin', 'Dest']].head()

Informació dels vols més llargs:



Distance UniqueCarrier  FlightNum TailNum Origin Dest
872155       4962            CO         15  N76055    EWR  HNL
360576       4962            CO         15  N67058    EWR  HNL
1923805      4962            CO         15  N78060    EWR  HNL
177271       4962            CO         15  N76054    EWR  HNL
1926817      4962            CO         15  N67058    EWR  HNL

Els vols més llargs són els que fa l'aerolinia CO de l'aeroport Internacional de Newark Liberty (Nova Jersey) a l'aeroport  Internacional Daniel K. Inouye (Honolulu).

#### I els més endarrerits? 

In [48]:
print("Informació dels vols més endarrerits:\n")

highest_delays = flights.sort_values(by='ArrDelay', ascending=False)
highest_delays[['ArrDelay', 'UniqueCarrier', 'FlightNum', 'TailNum', 'Origin', 'Dest']].head()

Informació dels vols més endarrerits:



ArrDelay UniqueCarrier  FlightNum TailNum Origin Dest
1498689    1382.0            NW        808  N808NW    HNL  MSP
1380381    1359.0            NW        624  N376NW    ABQ  MSP
1027010    1344.0            NW        776  N501US    LAS  MSP
1400101    1286.0            AA       1968  N4WLAA    SLC  DFW
1593181    1215.0            MQ       3560  N942MQ    GJT  DFW

Aquí veiem els vols més endarrerits. El major retard és de 1382 minuts.

#### Busca les rutes més llargues 

In [49]:
longest_routes = longest_flights[['Origin', 'Dest', 'Distance']]

# Ordenem les columnes 'Origin' i 'Dest' per evitar que les rutes en direcció inversa es considerin com a dues rutes diferents.
longest_routes[['Origin', 'Dest']] = longest_routes.apply(lambda row: sorted([row['Origin'], row['Dest']]), axis=1, result_type='expand')

longest_routes = longest_routes.drop_duplicates()

print('Rutes més llargues:\n')
longest_routes.head()

Rutes més llargues:



/var/folders/q_/v72nw5w576x1lnq29nqcs15r0000gn/T/ipykernel_70847/1896514714.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  longest_routes[['Origin', 'Dest']] = longest_routes.apply(lambda row: sorted([row['Origin'], row['Dest']]), axis=1, result_type='expand')


Origin Dest  Distance
872155     EWR  HNL      4962
1427945    ATL  HNL      4502
66596      HNL  ORD      4243
956585     KOA  ORD      4213
447762     OGG  ORD      4184

Tal i com podem veure les dues rutes més llargues tenen com a punt d'origen o destí Honolulu.

#### Busca les rutes que acumulen més retards¶

In [51]:
delays_per_route = highest_delays.groupby(['Origin', 'Dest'])['ArrDelay'].sum().reset_index()

highest_delays[['Origin', 'Dest']] = highest_delays.apply(lambda row: sorted([row['Origin'], row['Dest']]), axis=1, result_type='expand')

delays_per_route = delays_per_route.drop_duplicates().sort_values(by='ArrDelay', ascending=False)

print('Retards acumulats per ruta:\n')
delays_per_route.head()

Retards acumulats per ruta:



Origin Dest  ArrDelay
1944    LGA  ORD   47875.0
1919    LAX  SFO   41529.0
201     ATL  LGA   39816.0
1387    EWR  ORD   39740.0
162     ATL  EWR   38539.0

La ruta de l'aeroport LaGuardia (Nova York) a l'aeroport Internacional O'Hare (Chicago), en els dos sentits, és la que acumula més retards.

#### 2.6 Aporta allò que consideris rellevant.

In [52]:
print('Retards acumulats segons el dia de la setmana:\n')
      
delays_per_dow = flights.groupby('DayOfWeek')['ArrDelay'].sum().to_frame()
delays_per_dow.sort_values(by='ArrDelay', ascending=False)

Retards acumulats segons el dia de la setmana:



ArrDelay
DayOfWeek           
5          1457897.0
7          1331528.0
1          1258362.0
4          1225817.0
2          1190694.0
3          1088699.0
6           907371.0

El dia amb més retards acumulats és el divendres, seguit del diumenge, coincidint amb els dies en els que la gent sol viatjar durant el cap de setmana. El dia amb menys retards és el dissabte.

In [53]:
print('Contribució de cada causa al retard:\n')

flights_with_delays = flights[flights['Delay'] == 1]
delay_data = flights_with_delays[['ArrDelay','CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']]
delay_proportions = delay_data.div(delay_data['ArrDelay'], axis=0) * 100
delay_proportions.mean().to_frame().rename(columns={0: 'Delay_contribution'})

Contribució de cada causa al retard:



Delay_contribution
ArrDelay                   100.000000
CarrierDelay                32.700690
WeatherDelay                 4.458355
NASDelay                    22.646647
SecurityDelay                0.249430
LateAircraftDelay           39.944878

Les causes que més contribueixen al retard en els vols són: la propagació del retard (degut al retard en el vol anterior), amb casi un 40%, i el retard associat a l'aerolínia, amb un 32,7%. La seguretat és, amb diferència, la causa que menys contribueix al retard total, ramb menys d'un 1%.

In [54]:
print('Retards per aerolínia:\n')

delays_per_carrier = flights.groupby('UniqueCarrier')['Delay'].sum()
delays_per_carrier.sort_values(ascending = False).head().to_frame().rename(columns={'Delay': 'Total_Delays'})

Retards per aerolínia:



Total_Delays
UniqueCarrier              
WN                    21060
AA                    13688
MQ                    10122
UA                    10002
OO                     9182

Les 5 aerolínies amb més retards coincideixen amb les aerolinies amb un major retard acumulat (en minuts). La única diferència està en las aerolínies United Airlines (UA) i American Eagle Airlines (MQ), doncs l'aerolínia UA és la tercera amb més minuts de retard acumulats però la quarta amb més retards. En el cas de l'aerolínia MQ succeix el contrari.

Per tant, encara MQ té més retards que UA, la segona acumula més minuts de retard.

#### Exercici 3. Exporta el dataset net i amb les noves columnes a Excel.

In [55]:
flights.to_excel('/Users/juliacorominasfont/Downloads/airlines_data.xlsx', sheet_name='Delays')